test/jdk/java/net/Socket/asyncClose/Leaky.java

```diff
@@ -0,0 +1,230 @@
+    }
+
+    @AfterAll
+    public static void finish() throws Exception {
+        executor.close();
+        listener.close();
+    }
+
+    /**
+     * Race Socket bind and close.
+     */
+    @RepeatedTest(100)
+    public void raceBindAndClose() throws Exception {
+        Socket socket = new Socket();
+
+        race(socket::close, () -> {
+            try {
+                socket.bind(new InetSocketAddress(0));
+            } catch (IOException ioe) {
+                if (!socket.isClosed()) {
+                    throw ioe;
+                }
+            }
+        });
+
+        // check that there isn't an open socket
+        SocketImpl psi = getPlatformSocketImpl(socket);
+        assertFalse(isSocketOpen(psi));
+    }
+
+    /**
+     * Race Socket connect and close.
+     */
+    @RepeatedTest(100)
+    public void raceConnectAndClose() throws Exception {
+        Socket socket = new Socket();
+
+        race(socket::close, () -> {
+            try {
+                socket.connect(listener.getLocalSocketAddress());
+                // if connected, need to close other end
+                listener.accept().close();
+            } catch (IOException ioe) {
+                if (!socket.isClosed()) {
+                    throw ioe;
+                }
+            }
+        });
+
+        // check that there isn't an open socket
+        SocketImpl psi = getPlatformSocketImpl(socket);
+        assertFalse(isSocketOpen(psi));
+    }
+
+    /**
+     * Race Socket getOption and close.
+     */
+    @RepeatedTest(100)
+    public void raceGetOptionAndClose() throws Exception {
+        Socket socket = new Socket();
+
+        race(socket::close, () -> {
+            try {
+                socket.getOption(StandardSocketOptions.SO_REUSEADDR);
+            } catch (IOException ioe) {
+                if (!socket.isClosed()) {
+                    throw ioe;
+                }
+            }
+        });
+
+        // check that there isn't an open socket
+        SocketImpl psi = getPlatformSocketImpl(socket);
+        assertFalse(isSocketOpen(psi));
+    }
+
+    /**
+     * A task that may throw an exception.
+     */
+    private interface ThrowingRunnable {
+        void run() throws Exception;
+    }
+
+    /**
+     * Staggers two tasks to execute after random delays.
+     */
+    private void race(ThrowingRunnable task1, ThrowingRunnable task2) throws Exception {
+        int delay1 = ThreadLocalRandom.current().nextInt(10);
+        int delay2 = ThreadLocalRandom.current().nextInt(10);
+
+        Future<Void> future1 = executor.schedule(() -> {
+            task1.run();
+            return null;
+        }, delay1, TimeUnit.MILLISECONDS);
+
+        Future<Void> future2 = executor.schedule(() -> {
+            task2.run();
+            return null;
+        }, delay2, TimeUnit.MILLISECONDS);
+
+        try {
+            future1.get();
+        } finally {
+            future2.get();
```

Comment:
Should we capture/report any exception thrown from `future1.get()`? In its current form that might be lost if even `future2.get()` throws some other exception.

Reply:
That's a good idea, in the event that the test fails for some reason.